src/stage.ts

```diff
@@ -0,0 +1,68 @@
+import { Middleware, SceneContextOptions } from './types'
+import BaseScene from './scenes/base'
+import Composer from './composer'
+import SceneContext from './scenes/context'
+import TelegrafContext from './context'
+const { compose, optional, lazy, passThru } = Composer
+
+class Stage<TContext extends TelegrafContext> extends Composer<TContext> {
+  options: SceneContextOptions
+  scenes: Map<string, BaseScene<TContext>>
+  constructor(
+    scenes: Array<BaseScene<TContext>> = [],
+    options?: SceneContextOptions
+  ) {
+    super()
+    this.options = {
+      sessionName: 'session',
+      ...options,
+    }
+    this.scenes = new Map()
+    scenes.forEach((scene) => this.register(scene))
+  }
+
+  register(...scenes: Array<BaseScene<TContext>>) {
+    scenes.forEach((scene) => {
+      if (!scene || !scene.id || typeof scene.middleware !== 'function') {
+        throw new Error('telegraf: Unsupported scene')
+      }
+      this.scenes.set(scene.id, scene)
+    })
+    return this
+  }
+
+  middleware(): Middleware.Fn<TContext> {
+    const handler = compose([
+      (ctx: TContext, next) => {
+        ctx.scene = new SceneContext<TContext>(ctx, this.scenes, this.options)
+        return next()
+      },
+      super.middleware() as Middleware.Fn<TelegrafContext>,
+      lazy((ctx) => ctx.scene.current || passThru()),
+    ])
+    return optional(
+      (ctx: { [key: string]: any }) => ctx[this.options.sessionName],
+      handler
+    ) as Middleware.Fn<TContext>
```

Comment:
```suggestion
    return Composer.unwrap(
      Composer.optional(
        (ctx) => (ctx as any)[this.options.sessionName],
        handler
      )
    )
```

`Composer.unwrap` would be unnecessary if this was removed: https://github.com/telegraf/telegraf/blob/2e4a9d4ca45175c4de01e8a38706f4eb0bdd05cc/src/composer.ts#L242-L244